File: replication/src/test/com/db4o/test/replication/db4ounit/DrsTestCase.java
/* Copyright (C) 2004 - 2006 db4objects Inc. http://www.db4o.com */

package com.db4o.test.replication.db4ounit;

import java.util.List;
import java.util.Map;

import com.db4o.Db4o;
import com.db4o.ObjectSet;
import com.db4o.inside.replication.TestableReplicationProviderInside;
import com.db4o.replication.Replication;
import com.db4o.replication.ReplicationEventListener;
import com.db4o.replication.ReplicationSession;
import com.db4o.test.drs.Car;
import com.db4o.test.drs.CollectionHolder;
import com.db4o.test.drs.ListContent;
import com.db4o.test.drs.ListHolder;
import com.db4o.test.drs.MapContent;
import com.db4o.test.drs.MapHolder;
import com.db4o.test.drs.Pilot;
import com.db4o.test.drs.R0;
import com.db4o.test.drs.Replicated;
import com.db4o.test.drs.SPCChild;
import com.db4o.test.drs.SPCParent;
import com.db4o.test.drs.SimpleArrayContent;
import com.db4o.test.drs.SimpleArrayHolder;

import db4ounit.Assert;
import db4ounit.TestCase;
import db4ounit.TestLifeCycle;

public abstract class DrsTestCase implements TestCase, TestLifeCycle {
	
	public static final Class[] mappings;
	public static final Class[] extraMappingsForCleaning = new Class[]{Map.class, List.class};

	static {
		mappings = new Class[]{CollectionHolder.class, Replicated.class,
				SPCParent.class, SPCChild.class,
				ListHolder.class, ListContent.class,
				MapHolder.class, MapContent.class,
				SimpleArrayContent.class, SimpleArrayHolder.class,
				R0.class, Pilot.class, Car.class};
	}
	

	private DrsFixture _a;
	private DrsFixture _b;
	
	public void setUp() throws Exception {
		cleanBoth();
		configure();
		openBoth();
		store();
		reopen();
	}

	private void cleanBoth() {
		_a.clean();
		_b.clean();
	}

	protected void clean() {
		for (int i = 0; i < mappings.length; i++) {
			a().provider().deleteAllInstances(mappings[i]);
			b().provider().deleteAllInstances(mappings[i]);
		}

		for (int i = 0; i < extraMappingsForCleaning.length; i++) {
			a().provider().deleteAllInstances(extraMappingsForCleaning[i]);
			b().provider().deleteAllInstances(extraMappingsForCleaning[i]);
		}

		a().provider().commit();
		b().provider().commit();
	}

	protected void store() {}
	
	protected void configure() {
		Db4o.configure().generateUUIDs(Integer.MAX_VALUE);
		Db4o.configure().generateVersionNumbers(Integer.MAX_VALUE);
	}
	
	protected void reopen() throws Exception {
		closeBoth();
		openBoth();
	}

	private void openBoth() throws Exception {
		_a.open();
		_b.open();
	}
	
	public void tearDown() throws Exception {
		closeBoth();
		cleanBoth();
	}

	private void closeBoth() throws Exception {
		_a.close();
		_b.close();
	}
	
	public void a(DrsFixture fixture) {
		_a = fixture;
	}
	
	public void b(DrsFixture fixture) {
		_b = fixture;
	}
	
	public DrsFixture a() {
		return _a;
	}

	public DrsFixture b() {
		return _b;
	}

	protected void ensureOneInstance(TestableReplicationProviderInside provider, Class clazz) {
		ensureInstanceCount(provider, clazz, 1);
	}

	protected void ensureInstanceCount(TestableReplicationProviderInside provider, Class clazz, int count) {
		ObjectSet objectSet = provider.getStoredObjects(clazz);
		Assert.areEqual(count, objectSet.size());
	}

	protected Object getOneInstance(TestableReplicationProviderInside provider, Class clazz) {
		ObjectSet objectSet = provider.getStoredObjects(clazz);

		if (1 != objectSet.size())
			throw new RuntimeException("Found more than one instance of + " + clazz + " in provider = " + provider);

		return objectSet.next();
	}

	protected void replicateAll(TestableReplicationProviderInside providerFrom, TestableReplicationProviderInside providerTo) {
		//System.out.println("from = " + providerFrom + ", to = " + providerTo);
		ReplicationSession replication = Replication.begin(providerFrom, providerTo);
		ObjectSet allObjects = providerFrom.objectsChangedSinceLastReplication();

		if (!allObjects.hasNext())
			throw new RuntimeException("Can't find any objects to replicate");

		while (allObjects.hasNext()) {
			Object changed = allObjects.next();
			//System.out.println("changed = " + changed);
			replication.replicate(changed);
		}
		replication.commit();
	}
	
	protected void replicateAll(
			TestableReplicationProviderInside from, TestableReplicationProviderInside to, ReplicationEventListener listener) {
		ReplicationSession replication = Replication.begin(from, to, listener);

		ObjectSet allObjects = from.objectsChangedSinceLastReplication();
		while (allObjects.hasNext()) {
			Object changed = allObjects.next();
			//System.out.println("changed = " + changed);
			replication.replicate(changed);
		}

		replication.commit();
	}

	protected void delete(Class[] classes) {
		_a.clean();
/*		for (int i = 0; i < classes.length; i++) {
			_a.deleteAllInstances(classes[i]);
			_b.deleteAllInstances(classes[i]);
		}
		_a.commit();
		_b.commit(); */
	}

	protected void replicateClass(TestableReplicationProviderInside providerA, TestableReplicationProviderInside providerB, Class clazz) {
		//System.out.println("ReplicationTestcase.replicateClass");
		ReplicationSession replication = Replication.begin(providerA, providerB);
		ObjectSet allObjects = providerA.objectsChangedSinceLastReplication(clazz);
		while (allObjects.hasNext()) {
			final Object obj = allObjects.next();
			//System.out.println("obj = " + obj);
			replication.replicate(obj);
		}
		replication.commit();
	}

	protected static void sleep(int millis) {
		try {
			Thread.sleep(millis);
		} catch (InterruptedException e) {
			throw new RuntimeException(e);
		}
	}

}

File: drs/drs.net/drs.tests.in/com/db4o/drs/test/DrsTestCase.cs
namespace com.db4o.drs.test
{
    public abstract class DrsTestCase : Db4oUnit.TestCase, Db4oUnit.TestLifeCycle
    {
        public static readonly System.Type[] mappings;

        public static readonly System.Type[] extraMappingsForCleaning = new System.Type[] { typeof(System.Collections.IDictionary), typeof(System.Collections.IList) };

        static DrsTestCase()
        {
            mappings = new System.Type[] { typeof(SPCParent), 
                                           typeof(SPCChild), 
                                           typeof(SimpleArrayContent), 
                                           typeof(SimpleArrayHolder), 
                                           typeof(Pilot), 
                                           typeof(Car) };
        }

        private com.db4o.drs.test.DrsFixture _a;

        private com.db4o.drs.test.DrsFixture _b;

        public virtual void SetUp()
        {
            CleanBoth();
            Configure();
            OpenBoth();
            Store();
            Reopen();
        }

        private void CleanBoth()
        {
            _a.Clean();
            _b.Clean();
        }

        protected virtual void Clean()
        {
            for (int i = 0; i < mappings.Length; i++)
            {
                A().Provider().DeleteAllInstances(mappings[i]);
                B().Provider().DeleteAllInstances(mappings[i]);
            }
            for (int i = 0; i < extraMappingsForCleaning.Length; i++)
            {
                A().Provider().DeleteAllInstances(extraMappingsForCleaning[i]);
                B().Provider().DeleteAllInstances(extraMappingsForCleaning[i]);
            }
            A().Provider().Commit();
            B().Provider().Commit();
        }

        protected virtual void Store()
        {
        }

        protected virtual void Configure()
        {
            com.db4o.Db4o.Configure().GenerateUUIDs(int.MaxValue);
            com.db4o.Db4o.Configure().GenerateVersionNumbers(int.MaxValue);
        }

        protected virtual void Reopen()
        {
            CloseBoth();
            OpenBoth();
        }

        private void OpenBoth()
        {
            _a.Open();
            _b.Open();
        }

        public virtual void TearDown()
        {
            CloseBoth();
            CleanBoth();
        }

        private void CloseBoth()
        {
            _a.Close();
            _b.Close();
        }

        public virtual void A(com.db4o.drs.test.DrsFixture fixture)
        {
            _a = fixture;
        }

        public virtual void B(com.db4o.drs.test.DrsFixture fixture)
        {
            _b = fixture;
        }

        public virtual com.db4o.drs.test.DrsFixture A()
        {
            return _a;
        }

        public virtual com.db4o.drs.test.DrsFixture B()
        {
            return _b;
        }

        protected virtual void EnsureOneInstance(com.db4o.drs.inside.TestableReplicationProviderInside
             provider, System.Type clazz)
        {
            EnsureInstanceCount(provider, clazz, 1);
        }

        protected virtual void EnsureInstanceCount(com.db4o.drs.inside.TestableReplicationProviderInside
             provider, System.Type clazz, int count)
        {
            com.db4o.ObjectSet objectSet = provider.GetStoredObjects(clazz);
            Db4oUnit.Assert.AreEqual(count, objectSet.Size());
        }

        protected virtual object GetOneInstance(com.db4o.drs.inside.TestableReplicationProviderInside
             provider, System.Type clazz)
        {
            com.db4o.ObjectSet objectSet = provider.GetStoredObjects(clazz);
            if (1 != objectSet.Size())
            {
                throw new j4o.lang.RuntimeException("Found more than one instance of + " + clazz
                    + " in provider = " + provider);
            }
            return objectSet.Next();
        }

        protected virtual void ReplicateAll(com.db4o.drs.inside.TestableReplicationProviderInside
             providerFrom, com.db4o.drs.inside.TestableReplicationProviderInside providerTo)
        {
            com.db4o.drs.ReplicationSession replication = com.db4o.drs.Replication.Begin(providerFrom
                , providerTo);
            com.db4o.ObjectSet allObjects = providerFrom.ObjectsChangedSinceLastReplication();
            if (!allObjects.HasNext())
            {
                throw new j4o.lang.RuntimeException("Can't find any objects to replicate");
            }
            while (allObjects.HasNext())
            {
                object changed = allObjects.Next();
                replication.Replicate(changed);
            }
            replication.Commit();
        }

        protected virtual void ReplicateAll(com.db4o.drs.inside.TestableReplicationProviderInside
             from, com.db4o.drs.inside.TestableReplicationProviderInside to, com.db4o.drs.ReplicationEventListener
             listener)
        {
            com.db4o.drs.ReplicationSession replication = com.db4o.drs.Replication.Begin(from
                , to, listener);
            com.db4o.ObjectSet allObjects = from.ObjectsChangedSinceLastReplication();
            while (allObjects.HasNext())
            {
                object changed = allObjects.Next();
                replication.Replicate(changed);
            }
            replication.Commit();
        }

        protected virtual void Delete(System.Type[] classes)
        {
            for (int i = 0; i < classes.Length; i++)
            {
                A().Provider().DeleteAllInstances(classes[i]);
                B().Provider().DeleteAllInstances(classes[i]);
            }
            A().Provider().Commit();
            B().Provider().Commit();
        }

        protected virtual void ReplicateClass(com.db4o.drs.inside.TestableReplicationProviderInside
             providerA, com.db4o.drs.inside.TestableReplicationProviderInside providerB, System.Type
             clazz)
        {
            com.db4o.drs.ReplicationSession replication = com.db4o.drs.Replication.Begin(providerA
                , providerB);
            com.db4o.ObjectSet allObjects = providerA.ObjectsChangedSinceLastReplication(clazz
                );
            while (allObjects.HasNext())
            {
                object obj = allObjects.Next();
                replication.Replicate(obj);
            }
            replication.Commit();
        }

        protected static void Sleep(int millis)
        {
            try
            {
                j4o.lang.Thread.Sleep(millis);
            }
            catch (System.Exception e)
            {
                throw new j4o.lang.RuntimeException(e);
            }
        }
    }
}
